Add external callbacks into RecordFunction

ilia-cher opened the pull request:
Summary:
Add a way to insert external callbacks into PT's RecordFunction

Test Plan:
python setup.py develop --cmake
pytest -s -v test/test_autograd.py::TestAutograd::test_profiler

./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

Commit:
Add external callbacks into RecordFunction

Summary:
Add a way to insert external callbacks into PT's RecordFunction

Test Plan:
python setup.py develop --cmake

pytest -s -v test/test_autograd.py::TestAutograd::test_profiler

./build/bin/test_jit --gtest_filter=JitTest.RecordFunction

apaszke reviewed (changes requested):
Did we do any perf checks on this? This is highly performance sensitive code, and we had to fix perf regressions over and over.

apaszke commented on torch/csrc/autograd/profiler.h:
```diff
@@ -203,6 +203,65 @@ TORCH_API void mark(std::string name, bool include_cuda = true);
+
+ private:
+  Function* fn_ = nullptr;
+  std::string name_;
```
Having to keep and initialize an `std::string` per event might be a bit much (especially that this is a rarely used functionality). That's why we used to keep `std::unique_ptr<std::string>` an an Event (there's a comment explaining that too).

apaszke reviewed (commented):
Why do we need those callbacks?

ilia-cher commented:
@apaszke,
the reasons to have callbacks are: 1) plugging in external perf logging code and 2) potentially further unification of tracer and/or other components behind a single observer API;
I'll add results for perf testing, the goal is to have none or almost none overhead when not explicitly using extra callbacks

zdevito commented on torch/csrc/autograd/init.cpp:
```diff
@@ -50,6 +50,22 @@ PyObject * THPAutograd_initExtension(PyObject *_unused)
+    start_fn.inc_ref(); //(?)
+    end_fn.inc_ref();
+    extra_arg.inc_ref();
+    torch::autograd::profiler::RecordFunctionRegistry::registerCallbacks(
```
The formulation as two std::functions makes it hard for state to be passed from the entry of the recording block to the exit. It is not problem for the current use since it has a thread-local push/pop stack, but in other uses (e.g. if we unify this with the tracer) will have difficulty correlating a start_fn with an end_fn. I don't have a great solution for this though.

zdevito reviewed (commented):
I think we should look into the perf issues that @apaszke mentioned before merging this, since it is a pain when the profiler is not efficient and adds overhead to traces. We might want to consider a nicer way of passing state from the beginning of a section to an end, maybe:

```
struct State {
    virtual ~State() {};
};
using RecordFunctionCallback = std::function<std::unique_ptr<State>(FunctionCallContext* ctx)>;
void registerRecordCallback(const RecordFuntionCallback& callback);
```

ilia-cher commented on torch/csrc/autograd/init.cpp:
```diff
@@ -50,6 +50,22 @@ PyObject * THPAutograd_initExtension(PyObject *_unused)
 
+  // used for testing and debugging
+  m.def("_register_callbacks", [](py::function start_fn, py::function end_fn, py::object extra_arg) {
+    start_fn.inc_ref(); //(?)
```
this is just to simplify testing in python (and potentially could be used for an easy debug later), I'm not yet sure about correct usage of refs here though, still investigating

ilia-cher replied:
will use a cpp test instead

ilia-cher commented:
addressed some comments, refactored and simplified, using existing test testAutogradProfiler

also did some benchmarking, mainly using code from testAutogradProfiler:
```
diff --git a/test/cpp/jit/test_misc.h b/test/cpp/jit/test_misc.h
index 1c7456b1c..708d17624 100644
--- a/test/cpp/jit/test_misc.h
+++ b/test/cpp/jit/test_misc.h
@@ -1796,9 +1796,12 @@ void testAutogradProfiler() {
   std::stringstream ss;
   {
     autograd::profiler::RecordProfile guard(ss);
-    for (size_t i = 0; i < 100; ++i) {
+    auto before_time_ns = autograd::profiler::getTime();
+    for (size_t i = 0; i < 1000; ++i) {
       std::tie(hx, cx) = lstm(input[0], hx, cx, w_ih, w_hh);
     }
+    auto elapsed_time_ms = (autograd::profiler::getTime() - before_time_ns) / 10e6f;
+    std::cout << "Elapsed time: " << std::setw(5) << elapsed_time_ms << " ms." << std::endl;
   }

   std::string result = ss.str();
@@ -1806,7 +1809,7 @@ void testAutogradProfiler() {
   for (size_t pos = 0; (pos = result.find("tanh", pos)) != std::string::npos;
        count++, pos++) {
   }
-  AT_CHECK(count == 200);
+  AT_CHECK(count == 2000);
 }
```
N=100, mean/sd, ms.:
master: 27.0/1.3
change (w/ an extra pair of dummy callbacks): 27.2/1.6

ilia-cher commented:
will upd with a state passing

apaszke commented:
LSTM will contain a lot of code unrelated to the profiler, so we don't see the real impact of this change. In particular, when I'm investigating perf regressions I like to sprinkle `RecordFunction` over the places I suspect for slowness, which increases the number of events by an order of magnitude at least (e.g. log every single tensor destruction). It would be much better to have a benchmark that measures the overhead of a single `RecordFunction` call. Also, please use a test suite that runs for more than 30ms. This is definitely not enough time to reliably avoid system noise.

ilia-cher commented:
> LSTM will contain a lot of code unrelated to the profiler

This is the implementation used: https://github.com/pytorch/pytorch/blob/master/test/cpp/jit/test_misc.h#L342
It seems to be primarily tensor ops, which should be recorded by RecordFunction. What are the parts unrelated to the profiler?

> I like to sprinkle `RecordFunction` over the places that I suspect to be slow, which increases the number of events by an order of magnitude at least (e.g. log every single tensor destruction). It would be much better to have a benchmark that measures the overhead of a single `RecordFunction` call.

I'm not sure I follow, would you want to add an extra RecordFunction into a place where we do not have it in the code and measure the slowdown? What would be the workload you suggest we could use to measure runtime?

ilia-cher commented:
Using

```
index 1c7456b1c..0cfacd931 100644
--- a/test/cpp/jit/test_misc.h
+++ b/test/cpp/jit/test_misc.h
@@ -1807,6 +1807,15 @@ void testAutogradProfiler() {
        count++, pos++) {
   }
   AT_CHECK(count == 200);
+
+  auto x = torch::randn({1}, at::kCPU);
+  auto y = torch::randn({1}, at::kCPU);
+  auto before_time_ns = autograd::profiler::getTime();
+  for (size_t i = 0; i < 1000000; ++i) {
+    x.add_(y);
+  }
+  auto elapsed_time_ms = (autograd::profiler::getTime() - before_time_ns) / 10e6f;
+  std::cout << "Elapsed time: " << std::setw(5) << elapsed_time_ms << " ms." << std::endl;
 }

```

Results in the same runtimes before and after the change: (mean/sd before and after, ms: 145.8/3.4 vs 145.3/3.1)

apaszke commented:
@ilia-cher can you please point me to the line where you enable/disable the profiler in that test? It's obvious that there will be no difference if you don't turn it on. A quick search for `enableProfiler` in that file shows no hits.

ilia-cher commented:
@apaszke, I will add the version of a loop with profiler turned on, but first I'd want to share why I didn't turn it on in the test above. The reason is that this change does not change profiling code but only the plumbing around it (RecordFunction), so if we turn on profiling then on each iteration we will spend additional time doing profiling itself (e.g. adding events into thread local event lists), and that would render the overhead from RecordFunction even less noticeable.
I'll add the runtimes with profiler turned on too, and remove the loop from the test, accidentally left it there after debugging.

ilia-cher commented:
As expected, there's no statistically significant difference before vs after the change with the profiler turned on:
```
  {
    autograd::profiler::RecordProfile guard(ss);
    auto x = torch::randn({1}, at::kCPU);
    auto y = torch::randn({1}, at::kCPU);
    auto before_time_ns = autograd::profiler::getTime();
    for (size_t i = 0; i < 1000000; ++i) {
      x.add_(y);
    }
    auto elapsed_time_ms = (autograd::profiler::getTime() - before_time_ns) / 10e6f;
    std::cout << "Elapsed time: " << std::setw(5) << elapsed_time_ms << " ms." << std::endl;
  }
```
before, ms: 170.942, 162.889, 164.558, 160.912, 166.991
after, ms: 151.231, 169.231, 156.292, 159.606, 169.476

Also removed unneeded debug code

ilia-cher commented:
cc. @zdevito

apaszke commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -93,27 +93,126 @@ void popRange() {
+    const RecordFunctionCallback& start, const RecordFunctionCallback& end) {
+  std::lock_guard<std::mutex> guard(callbacks_mutex);
+  RecordCallbacks* new_callbacks = new RecordCallbacks;
+  RecordCallbacks* prev_callbacks = global_callbacks;
```
Why can't we use a `unique_ptr`?

apaszke commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -93,27 +93,126 @@ void popRange() {
+    Function* fn, UpdateCallContextCallback* ucb)
+    : callbacks_(global_callbacks) {
   pushRangeImpl(fn->name(), ", stashed seq=", fn->sequence_nr());
+  if (callbacks_) {
```
We never check that the profiler is actually on here? That seems weird

ilia-cher replied:
we don't necessarily want to check profiler flag for the callbacks

apaszke commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -93,27 +93,126 @@ void popRange() {
+
+namespace {
+std::mutex callbacks_mutex;
+std::atomic<RecordCallbacks*> global_callbacks(nullptr);
```
I don't get why does this have to be thread safe if the profiler in itself isn't? We can save on atomic load/stores and mutex locks if we get rid of this requirement.

ilia-cher replied:
because we'd want to be able to add/remove callbacks while the program is running

apaszke commented on torch/csrc/autograd/profiler.h:
```diff
@@ -203,18 +203,63 @@ TORCH_API void mark(std::string name, bool include_cuda = true);
+
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn);
+  explicit RecordFunction(Function* fn, UpdateCallContextCallback* ucb = nullptr);
```
What are `UpdateCallContextFallback`s for? Why can't they be just implemented as `RecordFunctionCallback`s? This seems really complicated.

ilia-cher replied:
UpdateCallContextCallback is used to pass extra data from the callsite and initialize additional fields of FunctionCallContext that are very expensive to do in general case

zdevito replied:
Even with this comment, I do not understand how UpdateCallContextCallback will be used. There doesn't seem to be any example uses in the tests either.

ilia-cher replied:
added an example (invokeTestRecordFunction) into test

apaszke reviewed (commented):
I still don't fully get why do we need those. Can't the external perf monitoring code sample traces? Why does it need to have access to live callbacks?

zdevito reviewed (commented):
See inline comments -- there are some issues around clarify of the implementation and correctness of the implementation.

ilia-cher commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -93,27 +93,126 @@ void popRange() {
+  new_callbacks->end_callbacks.push_back(end);
+  global_callbacks = new_callbacks;
+  if (prev_callbacks) {
+    delete prev_callbacks;
```
good catch!

ilia-cher commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -93,27 +93,126 @@ void popRange() {
+  if (!callbacks_) {
+    pushRangeImpl(std::move(name));
+  } else {
+    pushRangeImpl(name); // copy name
```
we could do this, I wasn't planning to do this in this change though

ilia-cher commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -93,27 +93,126 @@ void popRange() {
+RecordFunction::RecordFunction(
+    std::string name, UpdateCallContextCallback* ucb)
+    : callbacks_(global_callbacks) {
+  if (!callbacks_) {
```
tbh inlining was intentional, but I don't think its necessary, will update

ilia-cher commented:
Fixed issue with callbacks delete

ilia-cher commented:
Simplfied further - make pushRangeImpl/popRange callbacks too

ilia-cher commented:
Added a test that demonstrates how UpdateCallContextCallback can be used with RecordFunction to record input sizes

ilia-cher commented:
Using the same benchmark as above (ms)
before:168.468, 170.602, 172.813, 171.534, 175.609
after:168.805, 171.368, 166.192, 156.786, 176.363

apaszke commented:
The more I think about it the more I'm ambivalent about this test. Things that are the most annoying for me:
1. We use the infrastructure of the profiler, while the operation of callbacks is *completely independent* of the profiler. You never even check if it's enabled. That's a good sign that they should be separate components, and shouldn't live right next to each other.
2. This doesn't add any useful feature to our users. It's needed for some kind of an internal FB use-case, which is not generally exposed to others. We should avoid putting such code into the library, because FB is not the only user. If you see it as a feature others might like, I'd like to see a more specific pitch.

ilia-cher commented:
1. I'm not sure I follow this, the profiler is implemented as a callback, e.g. if we want to enable profiler we add a callback, if we want to disable it, we remove a callback, so it seems that the operation of callbacks is _not_ completely independent from the profiler?
2. RecordFunction becomes a generalized solution to insert callback code, that can be used not only for the profiling. For example, one important use case is that we can further add (autograd) tracer as a callback. This way we merge profiling and autograd tracing into one conceptually simple construct that's easy to understand, instead of relying to autogeneration of RecordFunction _and_ tracing code.

apaszke commented on torch/csrc/autograd/profiler.h:
```diff
@@ -203,23 +203,89 @@ TORCH_API void mark(std::string name, bool include_cuda = true);
+  std::unique_ptr<std::string> owned_name_;
+  const char* name_ptr_ = nullptr;
+  int64_t sequence_nr_ = -1;
+  std::unique_ptr<TensorSizesList> input_sizes;
```
If we want to implement tracing on top of this we'll need exact inputs (probably packed as `IValue`s), not only their sizes.

ilia-cher replied:
Yes, I agree, will remove the generalized UpdateFunctionCallContext and replace it with something more specific, it's main use case is inputs. Sometimes we have inputs ready as a stack of IValues (in JIT interpreter) but in VariableType we just have a codegen'd list of function arguments.

apaszke reviewed (commented):
Ok, I really like the current implementation (provided it keeps the expected performance level of our profiler because I haven't seen benchmarks since then), but if that's the plan then:
1. It should still live separately from the profiler, because it's much more general functionality
2. I really don't like `UpdateFunctionCallContext`. We should just pack the inputs into IValues (perhaps lazily??) and let the callbacks use that if they want to. It makes the API unnecessarily complicated.

ilia-cher commented:
@apaszke Yes, I agree, I will remove the generalized UpdateFunctionCallContext and replace it with something more specific, it's main use case is inputs. Sometimes we have inputs ready as a stack of IValues (in JIT interpreter) but in VariableType we just have a codegen'd list of function arguments.

ilia-cher commented on c10/util/C++17.h:
```diff
@@ -72,7 +72,7 @@ constexpr inline T&& forward(guts::remove_reference_t<T>&& t) noexcept {
 typename std::enable_if<!std::is_array<T>::value, std::unique_ptr<T>>::type
 make_unique(Args&&... args) {
-  return std::unique_ptr<T>(new T(forward<Args>(args)...));
+  return std::unique_ptr<T>(new T(c10::guts::forward<Args>(args)...));
```
these changes will be removed before merging, fixed as part of https://github.com/pytorch/pytorch/pull/18492

ilia-cher commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -0,0 +1,91 @@
+    const char* name, int64_t sequence_nr, GetPackedInputsCallback cb) {
+  if (!has_callbacks) {
+    return;
+  }
```
(will remove)

ilia-cher commented:
For the following benchmark:
```
  {
    autograd::profiler::RecordProfile guard(ss);
    auto x = torch::randn({1}, at::kCPU);
    auto y = torch::randn({1}, at::kCPU);
    auto before_time_ns = autograd::profiler::getTime();
    for (size_t i = 0; i < 1000000; ++i) {
      x.add_(y);
    }
    auto elapsed_time_ms = (autograd::profiler::getTime() - before_time_ns) / 10e6f;
    std::cout << "Elapsed time: " << std::setw(5) << elapsed_time_ms << " ms." << std::endl;
  }
```
before (ms): 171.037, 173.046, 175.663, 167.404, 194.821, 183.806, 175.031, 178.954, 156.267, 167.208, mean = 174.32
after (ms): 183.246, 186.265, 166.657, 182.19, 186.236, 179.199, 182.841, 186.07, 184.949, 184.328, mean = 182.2

ilia-cher commented:
Confirmed that dynamic allocation of callbacks in RecordFunction is the cause, if we'd want to have no difference in speed, then we can go with the same state class for all callbacks (and allocate it locally with SmallVector) or not pass state in callbacks at all (confirmed that there's almost no difference in speed in this case)

Commit:
No callback state version

Summary:

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

ilia-cher commented:
Added the second commit - a version with no callback state passed between start and end events.

ilia-cher commented:
If we don't use callback state:
178.648, 173.523, 173.572, 181.487, 178.418, 168.214, 165.498,  171.42, 172.774, 179.514 ms, mean 174.31, same speed as trunk

zdevito commented on torch/csrc/autograd/record_function.h:
```diff
@@ -0,0 +1,109 @@
+
+using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
+
+struct FunctionCallContext {
```
I am not sure we are gaining much from separating this from RecordFunction. It causes more allocations and refcounting to support recording through asynchronous calls, but it is not clear to me we really should be recording async trace information through forks anyway. If we have a different hook for profiling the interpreter anyway, then it would make more sense for that facility to be able to collect accurate interpreter stack traces.

ilia-cher replied:
One specific use case I had in mind - associating an op with a specific script. Interpreter knows what code it runs, while RecordFunction does the (sampled) logging and doesn't have the interpreter context.
If I understand your suggestion correctly, technically we could set the "starting" thread local logging info for the new thread (copied from the parent thread) and use raw pointers to link nested RecordFunctions.

ilia-cher replied:
That would remove necessity for refcounting and we could use c10::optional<FunctionCallContext> object as a member of RecordFunction. I guess from the design point of view one argument to keep FunctionCallContext is to control information passed to the callbacks by aggregating it in FunctionCallContext.

zdevito reviewed (commented):
Good progress. I still think we should keep the first version of this simple, and avoid refcounted allocations to hold state when we can just have a chain on the stack. Otherwise I just have minor comments.

ilia-cher commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -0,0 +1,98 @@
+// fn->name() ensures that nvtx annotations for custom function backward() methods
+// receive a relevant, demangled name.
+FunctionCallContext::FunctionCallContext(Function* fn, GetPackedInputsCallback cb)
+    : fn_(fn), owned_name_(new std::string(fn->name())), name_ptr_(owned_name_->c_str()),
```
Original reason was to save on FunctionCallContext size, another point is that if we want to keep a simple interface (const char* name()) then we'll need to store temporary string returned from fn->name(). Having said that, I do agree that in most cases there will be at most one active callback querying name() so we could optimize this further for the Function use case.

Commit:
comments

zdevito commented on torch/csrc/autograd/profiler.h:
```diff
@@ -97,13 +98,8 @@ enum class EventKind : uint16_t {
-  , thread_id_(thread_id) { record(record_cuda); }
-  Event(EventKind kind, const char* name, uint16_t thread_id, bool record_cuda)
-  : name_ptr_(name)
+  Event(EventKind kind, const StringView& name, uint16_t thread_id, bool record_cuda)
```
nit: it is possible to avoid the refcount:
StringView name,
... name_(std::move(name))

zdevito commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -0,0 +1,83 @@
+namespace torch { namespace autograd { namespace profiler {
+
+namespace {
+std::atomic<bool> has_callbacks(false);
```
nit: no need for atomic here since the rest isn't threadsafe

zdevito reviewed (approved):
Looks good! Only have some minor nits.

ilia-cher commented:
some nits

apaszke commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -34,95 +34,66 @@ RangeEventList& getEventList() {
-
-template<typename T>
-void pushRangeImpl(T name, const char* msg="", int64_t sequence_nr=-1) {
-  if (state == ProfilerState::Disabled) {
```
Why did we loose checks for the profiler state everywhere??? Those functions are exposed to Python, so they really need to check if the profiler is enabled or not!

ilia-cher replied:
sent https://github.com/pytorch/pytorch/pull/18908

apaszke commented on torch/csrc/autograd/profiler.cpp:
```diff
@@ -34,95 +34,66 @@ RangeEventList& getEventList() {
     getEventList().record(
         EventKind::PushRange,
-        std::move(name),
+        name,
```
You've added an unnecessary copy that didn't exist before

ilia-cher replied:
right, see my comment below

apaszke commented on torch/csrc/autograd/record_function.h:
```diff
@@ -0,0 +1,102 @@
+ private:
+  void processCallbacks();
+
+  Function* fn_ = nullptr;
```
Why do we keep this around?

apaszke commented on torch/csrc/autograd/record_function.h:
```diff
@@ -0,0 +1,102 @@
+  GetPackedInputsCallback inputs_cb_ = nullptr;
+  mutable bool inputs_initialized_ = false;
+  // initialized lazily by inputs_cb_
+  mutable std::vector<c10::IValue> inputs_;
```
Uh, why is this mutable?

ilia-cher replied:
I believe you're commenting an older version, please check the newer one

apaszke commented on torch/csrc/autograd/record_function.h:
```diff
@@ -0,0 +1,102 @@
+  mutable bool inputs_initialized_ = false;
+  // initialized lazily by inputs_cb_
+  mutable std::vector<c10::IValue> inputs_;
+};
```
I'm a bit concerned that previously `RecordFunction` was a trivial struct that didn't need any storage, while now it starts to consume considerable stack space. Why do we keep things like `sequence_nr_` or `name_` inside? Those should just get pushed into the event list and be forgotten.

ilia-cher replied:
we keep name_ inside because Event is not the only consumer, there might be other callbacks that need to have access to the function name

apaszke commented on torch/csrc/autograd/record_function.cpp:
```diff
@@ -0,0 +1,83 @@
+
+void RecordFunction::processCallbacks() {
+  parent_ = thread_local_func_;
+  thread_local_func_ = this;
```
Why do we even maintain this? It doesn't seem to be used anywhere.

ilia-cher replied:
this allows you to check what nested RecordFunction you are in, I've added an example in the test how it is used

apaszke replied:
Can you point me to the exact lines? When I search for this identifier in this patch then the only usages I can see are inside this file.

ilia-cher replied:
please check the `getFullName` function in `test_misc.h`

swolchok commented on torch/csrc/autograd/record_function.h:
```diff
@@ -0,0 +1,102 @@
+      const char* name,
+      GetPackedInputsCallback cb) : RecordFunction(name, -1, cb) {}
+
+  virtual ~RecordFunction();
```
why is this virtual?
